codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketProtocolHandler.java

```diff
@@ -61,8 +61,11 @@
     @Override
     protected void decode(ChannelHandlerContext ctx, WebSocketFrame frame) throws Exception {
         if (frame instanceof PingWebSocketFrame) {
-            frame.content().retain();
-            ctx.channel().writeAndFlush(new PongWebSocketFrame(frame.content()));
+            // We need to `send` the binary data to the pong frame, because the MessageToMessageDecoder
+            // is going to close the ping frame, which would otherwise cause the data to be freed.
+            try (frame) {
+                ctx.channel().writeAndFlush(new PongWebSocketFrame(frame.binaryData().send()));
```

Comment:
I think this is more correct (while not really related to this change tho).
```suggestion
                ctx.writeAndFlush(new PongWebSocketFrame(frame.binaryData().send()));
```